### PR TITLE
Add tree-level metric calculation

### PR DESCRIPTION
**Pull request by amrithasp02**
Added a script to calculate the tree-level accuracy values. This metric is calculated at the end of every validation loop. It requires a single metadata file containing ['image_id', 'tree_unique_id', 'species_final'] columns for all the validation set images. This file is created by https://github.com/open-forest-observatory/tree-species-prediction/blob/main/1_data_prep/13_tree_crops.py

There is an additional step to merge the metadata files which needs to go into the tree species repo -- I'll be adding that soon.

### CODE REVIEW

File: mmpretrain/evaluation/metrics/tree_metric.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+        # Map each tree_unique_id -> ground-truth label index
+        self.tree2label = {
+            row['tree_unique_id']: self.class_to_idx[row['species_final']]
+            for _, row in df.iterrows()
+        }
+
+        self.results = []
+
+    def process(self, data_batch, data_samples):
```
Comment (russelldj) [inline mmpretrain/evaluation/metrics/tree_metric.py:42]: Is `data_batch` the data used to generate the predictions? Since it's unused in this it might be worth just noting what it is.

File: mmpretrain/evaluation/metrics/tree_metric.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        
+        Args:
+            results (list[dict]): The processed results of each batch.
+
+        Returns:
+            dict: Dictionary with tree-level accuracy as {'tree_acc': float}.
+        """
+
+        # For every tree (key) append predictions from all of its images to a single list (value)
+        tree_preds = {}
```
Comment (russelldj) [inline mmpretrain/evaluation/metrics/tree_metric.py]: nit: I like using a `DefaultDict` initialized to an empty list to avoid a special case for the result not being there.

File: mmpretrain/evaluation/metrics/tree_metric.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                tree_preds[tid] = []
+            tree_preds[tid].append(r['pred'])
+
+        # Compute tree-level accuracy
+        correct = 0
+        total = 0
+        for tid, preds in tree_preds.items():
+            # Aggregate predictions across all images of the same tree
+            # Take mean of predictions and assign label with highest mean score
+            mean_pred = np.mean(preds, axis=0)
```
Comment (russelldj) [inline mmpretrain/evaluation/metrics/tree_metric.py:88]: Is this doing a confidence-weighted average? If so, maybe explain that.

File: mmpretrain/evaluation/metrics/tree_metric.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    def process(self, data_batch, data_samples):
+        """Collect per-image predictions."""
+        for sample in data_samples:
+            img_path = sample['img_path']
+            img_id = img_path.split('/')[-1].split('.')[0]  # filename without extension
+            # Convert prediction tensor to numpy array
+            pred = sample['pred_score'].cpu().numpy()
+
+            # Append prediction record with tree association
+            self.results.append({
```
Comment (russelldj) [inline mmpretrain/evaluation/metrics/tree_metric.py:59]: Does `self.results` get reset every epoch and/or a new object is created each epoch? I wouldn't want a continual average of all predictions across epochs.

**Review by russelldj — APPROVED** [review]
Looks good! Potentially a few things to update based on the questions, but feel free to merge it with whatever changes.

File: mmpretrain/evaluation/metrics/tree_metric.py (hunk 3 above)
Comment (amrithasp02) [reply mmpretrain/evaluation/metrics/tree_metric.py:88]: This is currently not doing confidence-weighted averaging. It's just taking the simple mean of softmax probabilities across all images and choosing the highest mean probability class. I think this would be something to experiment in the future. I've added a todo for now

File: mmpretrain/evaluation/metrics/tree_metric.py (hunk 4 above)
Comment (amrithasp02) [reply mmpretrain/evaluation/metrics/tree_metric.py:59]: Yes, `self.results` gets reset after each epoch in the base class `BaseMetric` by the `evaluate()` method after calling `compute_metrics()`
